Fix loading indicator and loading bar issues

MV-GH opened the pull request:
Fixes #773 
which prevents the loading indicator from showing when unwanted, also has fixes for loading indicator disappearing (as in hiding behind the the topbar) and overlay with posts.
Fixes loading bar not working in inbox and personprofileposts when loading in more posts/comments.

Also proposes #807 resetPosts is small relic from it.

MV-GH commented on app/src/main/java/com/jerboa/ui/components/home/HomeViewModel.kt:
```diff
@@ -42,7 +44,8 @@ class HomeViewModel : ViewModel() {
     var fetchingMore by mutableStateOf(false)
         private set
-
+    var refreshing by mutableStateOf(false)
```
The refreshindicator should only show if you pull down, every other action that causes it to load the posts (such as login, the more options refresh button) also trigger the ApiState.loading thus i cant use that, so I added a bunch of refreshing states that only track the state of when it is triggered by the user pulling down the indicator

MV-GH replied:
But getPosts and the others then set it back to loading

dessalines replied:
Then you could make another viewmodel function that doesn't set loading, or make it optional in some way.

Either way it makes sense to have a single ApiState object reflect what's going on, rather than multiple approaches on top of each other.

MV-GH replied:
I'm not following with `that makes loading optional`. I'm just doing a approach that's already been done, it's same as fetchingMore. Loading is still used for the loading bar. Adding a Apistate imo is much more complexity because now each exhaustive "when" structure will have to be updated to have this state that's completely irrelevant for those causes. To the api refreshing is irrelevant it doesn't need to differentiate between refreshing an action that's only done when a user pulls down an indicator vs when its waiting for a result. I can understand the onCompletion callback if you have another way, I'm happy to change it. But i don't understand why against the refreshing state. The viewmodels are littered with states already.

dessalines replied:
I really hated added fetchingMore, but it was unfortunately necessary for infinite scrolling. Ideally there shouldn't be any states besides ApiState.

I spose I'm okay though with this refreshing state, although I see this getting complicated very quickly with multiple loading and conflicting Api states.

MV-GH replied:
I think its actually possible to remove fetchingMore, and I see issue with it too that I can fix. If fetching more fails it will still skip to the next page and on next try it will skip loading that page and loads the next.

nahwneeth replied:
I too noticed that it's skipping pages when failing. How about making `fetchingMore` an ApiState of the current page and `postRes` just plain list? The `postRes` can be used directly in the lazy list without matching. The `fetchingMore` can be used to show loading state & possibly retry on error. If `fetchingMore` is loading & page is 1, it's refreshing and `postRes` has to be reset. Otherwise it's just loading and `postRes` will be extended with the result. Also, when incrementing pages if the previous page ended in failure (ie `fetchingMore` is ApiState.Failure) then retry the page instead. However, I think this might be a lot more work.

MV-GH replied:
I have already a working change that removes fetchingmore, ` loading & page is 1` this still doesn't guarantee that it only shows the indicator when you pull it down, for example you can use the refresh button in the more options on the first page and then it shouldnt show the indicator. And i have already included the fix fort he failing pages ` then retry the page instead` exactly what I am doing

MV-GH commented on app/src/main/java/com/jerboa/ui/components/inbox/InboxViewModel.kt:
```diff
@@ -97,8 +100,8 @@ class InboxViewModel : ViewModel() {
-    ) {
-        viewModelScope.launch {
+    ): Job {
+        return viewModelScope.launch {
```
I don't know what you are referring to ? the structure? I made it return the  Job because i need to do action on when this coroutine completes

dessalines replied:
Okay I see now.

dessalines reviewed (commented):
I should be able to test this tomorrow.

MV-GH commented:
@dessalines  I have a prototype for removing fetchingMore in homeCommunity, everything still works as before. Had to some changes though.

 I added ImmutableList because List is unstable see https://issuetracker.google.com/issues/199496149. (In essence even tho List is immutable you cant guarantee it always is because you can cast it to mutable) The library i added has lists that is considered stable. I moved the composable out of that when structure because it kept recreating it because i was changing the state of res which kept it lauching the scrolledToEnd() func which caused a bit of loop on loading comments but moving it out and using that stable list solved that.
 
 If you find this good I can apply this too all the fetchingMore

MV-GH commented:
I have fully removed fetchingMore, also included a bunch of fixes for small issues that i encountered ( topbar disappearing during load, missing resetpage)

I also changed ApiState to include an action that has a value while it is loading, called it Awaiting. Succes, Awaiting and potential other future apiStates that will have a value should inherit from Holder.

I was thinking of maybe changing Loading to Loading : Awaiting<Unit>(Unit) but instead I added a isLoading() extension.

Names can be changed if you have a better name for them.

dessalines commented on app/src/main/java/com/jerboa/ui/components/inbox/InboxViewModel.kt:
```diff
@@ -80,17 +79,29 @@ class InboxViewModel : ViewModel(), Initializable {
         mutableStateOf(ApiState.Empty)
 
-    var page by mutableStateOf(1)
+    var pageReplies by mutableStateOf(1)
```
Thanks for splitting these out.

dessalines commented on app/src/main/java/com/jerboa/ui/components/inbox/InboxViewModel.kt:
```diff
@@ -99,127 +110,133 @@ class InboxViewModel : ViewModel(), Initializable {
-                        }
+            val oldRes = repliesRes
+            when (oldRes) {
+                is ApiState.Success -> repliesRes = ApiState.Awaiting(oldRes.data)
```
I see now.

dessalines commented on app/src/main/java/com/jerboa/api/Http.kt:
```diff
@@ -316,9 +316,12 @@ interface API {
+
+    abstract class Holder<T>(val data: T) : ApiState<T>()
+    class Success<T>(data: T) : Holder<T>(data)
+    class Awaiting<T>(data: T) : Holder<T>(data) // TODO: need better name for this
```
This awaiting always seems to be used for fetching more data / Appending. Might as well call it `Appending`.

dessalines commented on app/src/main/java/com/jerboa/api/Http.kt:
```diff
@@ -316,9 +316,12 @@ interface API {
+    class Awaiting<T>(data: T) : Holder<T>(data) // TODO: need better name for this
     class Failure(val msg: Throwable) : ApiState<Nothing>()
-    object Loading : ApiState<Nothing>()
+    object Loading : ApiState<Nothing>() // TODO: maybe object Loading : Awaiting<Unit>(Unit)
```
You can remove this comment.

dessalines commented on app/src/main/java/com/jerboa/api/Http.kt:
```diff
@@ -316,9 +316,12 @@ interface API {
 
 sealed class ApiState<out T> {
-    class Success<T>(val data: T) : ApiState<T>()
+
```
I still don't fully understand also why you can't add a `Refreshing` state here, instead of having isRefreshing in all the viewmodels.

dessalines commented:
Since `List` is unstable, I'll add a ticket to convert all lists to ImmutableList.

dessalines commented:
I would, that would def be preferable, thx.

MV-GH commented:
Done

dessalines reviewed (approved):
Mmmk I just tested this, works great. I def appreciate that all the state is handled by `ApiState` now.
